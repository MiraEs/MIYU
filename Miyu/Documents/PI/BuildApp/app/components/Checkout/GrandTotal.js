'use strict';
import React, {
	Component,
} from 'react';
import PropTypes from 'prop-types';
import {
	StyleSheet,
	View,
} from 'react-native';
import {
	Text,
} from 'BuildLibrary';
import styles from '../../lib/styles';
import helpers from '../../lib/helpers';

const componentStyles = StyleSheet.create({
	grandTotal: {
		backgroundColor: styles.colors.white,
		borderColor: styles.colors.grey,
		borderTopWidth: styles.dimensions.borderWidth,
		paddingHorizontal: styles.measurements.gridSpace2,
		paddingVertical: styles.measurements.gridSpace1,
	},
	grandTotalRow: {
		flexDirection: 'row',
		alignItems: 'flex-end',
	},
});

class GrandTotal extends Component {

	calcGrandTotal = () => {
		const { cart, storeCredit, useStoreCredit } = this.props;
		const credit = useStoreCredit ? storeCredit : 0;
		const grandTotal = helpers.calcGrandTotal(cart, credit);

		return helpers.toUSD(grandTotal);
	};

	render() {

		return (
			<View style={componentStyles.grandTotal}>
				<View
					accessibilityLabel="Total Price"
					style={componentStyles.grandTotalRow}
				>
					<Text
						size="large"
						weight="bold"
						lineHeight={false}
						style={styles.elements.flex}
					>
						Grand Total
					</Text>
					<Text
						size="large"
						weight="bold"
						lineHeight={false}
						color="primary"
					>
						{this.calcGrandTotal()}
					</Text>
				</View>
			</View>
		);
	}
}

GrandTotal.propTypes = {
	cart: PropTypes.object,
	storeCredit: PropTypes.number,
	useStoreCredit: PropTypes.bool,
};

GrandTotal.defaultProps = {
	storeCredit: 0,
	useStoreCredit: false,
};

export default GrandTotal;
