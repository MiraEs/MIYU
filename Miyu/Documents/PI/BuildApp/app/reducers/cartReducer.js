import {
	GET_SESSION_CART_SUCCESS,
	UPDATE_SESSION_CART_SUCCESS,
	DELETE_SESSION_CART_SUCCESS,
	DELETE_SESSION_CART_ITEM_SUCCESS,
	GET_CUSTOMER_CARTS_SUCCESS,
	SET_SESSION_CART_ITEM_DELETE_STATUS,
	CLEAR_SESSION_CART_ITEM_DELETE_STATUS,
	SET_SESSION_CART_ITEM_PROPS,
	SET_SELECTED_SHIPPING_INDEX,
	ADD_COUPON_SUCCESS,
	REMOVE_COUPON_SUCCESS,
	CLEAR_SESSION_CART_SUCCESS,
	LOADING_SESSION_CART,
	UNDO_STATUS,
	COPY_SESSION_CART_SUCCESS,
	LOAD_QUOTE_SUCCESS,
	UPDATE_CART_ITEM_BOUNCE,
	GET_DELIVERY_DATES,
	GET_DELIVERY_DATES_SUCCESS,
	SET_DELIVERY_DATE,
	GET_SESSION_CART_ERRORS,
	SET_SESSION_CART_ERRORS,
} from '../constants/CartConstants';
export const initialState = {
	cart: {
		sessionCartItems: [],
		sessionCartErrors: [],
		checkingErrors: false,
	},
	carts: [],
	selectedShippingIndex: 0,
	sessionCartItemDeleteQueue: [],
	quoteId: null,
	isLoading: false,
	bounceFirstRowOnMount: false,
};

export function updateCartItem(cartItem, cartItems) {

	return cartItems.map((currentItem) => {
		let item = { ...currentItem };

		if (item.itemKey === cartItem.itemKey) {
			item = Object.assign({}, item, cartItem);
			item.hasChanged = true;
		}

		if (item.hasSubItems) {
			item.subItems = item.subItems.map((subItem) => {
				if (subItem.itemKey === cartItem.itemKey) {
					subItem = Object.assign({}, subItem, cartItem);
					item.hasChanged = true;
				}
				return subItem;
			});
		}
		return item;
	});
}

function cartReducer(state = initialState, action = {}) {
	const { cartItem, deleteId, itemKey } = action.payload || {};
	let cart, sessionCartItems;

	switch (action.type) {
		case COPY_SESSION_CART_SUCCESS:
			cart = {
				...action.payload,
				sessionCartItems: action.payload.sessionCartItems || [],
			};
			return {
				...state,
				isLoading: false,
				cart,
			};
		case GET_SESSION_CART_SUCCESS:
			sessionCartItems = action.cart.sessionCartItems || [];

			state.sessionCartItemDeleteQueue.forEach((queueItem) => {
				sessionCartItems.forEach((cartItem) => {
					if (cartItem.hasSubItems) {
						cartItem.subItems.forEach((subItem) => {
							if (subItem.itemKey === queueItem.itemKey) {
								subItem.deleteStatus = UNDO_STATUS;
							}
						});
					}

					if (cartItem.itemKey === queueItem.itemKey) {
						cartItem.deleteStatus = UNDO_STATUS;
					}
				});
			});

			cart = {
				...action.cart,
				sessionCartItems,
			};

			return {
				...state,
				isLoading: false,
				cart,
			};
		case UPDATE_SESSION_CART_SUCCESS:
			cart = {
				...action.cart,
				sessionCartItems: action.cart.sessionCartItems || [],
			};

			return {
				...state,
				isLoading: false,
				selectedShippingIndex: !action.zipCode ? 0 : state.selectedShippingIndex,
				cart,
			};
		case DELETE_SESSION_CART_SUCCESS:
			return {
				...state,
				isLoading: false,
				cart: {
					sessionCartItems: [],
				},
			};
		case DELETE_SESSION_CART_ITEM_SUCCESS:
			return {
				...state,
				isLoading: false,
				sessionCartItemDeleteQueue: state.sessionCartItemDeleteQueue.filter((item) => item.itemKey !== itemKey),
			};
		case GET_CUSTOMER_CARTS_SUCCESS:
			return {
				...state,
				isLoading: false,
				carts: action.carts.sort((a, b) => {
					if (a.created < b.created) {
						return 1;
					}
					return -1;
				}),
			};
		case SET_SESSION_CART_ITEM_DELETE_STATUS:
			cartItem.deleteStatus = UNDO_STATUS;
			cart = {
				...state.cart,
				sessionCartItems: updateCartItem(cartItem, state.cart.sessionCartItems),
			};

			return {
				...state,
				isLoading: false,
				sessionCartItemDeleteQueue: state.sessionCartItemDeleteQueue.concat({
					itemKey: cartItem.itemKey,
					status: UNDO_STATUS,
					deleteId,
				}),
				cart,
			};
		case CLEAR_SESSION_CART_ITEM_DELETE_STATUS:
			delete cartItem.deleteStatus;
			cart = {
				...state.cart,
				sessionCartItems: updateCartItem(cartItem, state.cart.sessionCartItems),
			};

			return {
				...state,
				isLoading: false,
				sessionCartItemDeleteQueue: state.sessionCartItemDeleteQueue.filter((item) => {
					if (item.itemKey === cartItem.itemKey) {
						clearTimeout(item.deleteId);
						return false;
					}
					return true;
				}),
				cart,
			};
		case SET_SESSION_CART_ITEM_PROPS:
			cart = {
				...state.cart,
				sessionCartItems: state.cart.sessionCartItems.map((item) => {
					if (item.itemKey === action.cartItem.itemKey) {
						item = Object.assign({}, item, action.props);
					}
					return item;
				}),
			};

			return {
				...state,
				isLoading: false,
				cart,
			};
		case SET_SELECTED_SHIPPING_INDEX:
			return {
				...state,
				isLoading: false,
				selectedShippingIndex: action.selectedShippingIndex > -1 ? action.selectedShippingIndex : 0,
			};
		case ADD_COUPON_SUCCESS:
			return {
				...state,
				isLoading: false,
				cart: action.cart,
			};
		case REMOVE_COUPON_SUCCESS:
			return {
				...state,
				isLoading: false,
				cart: action.payload,
			};
		case CLEAR_SESSION_CART_SUCCESS:
			return {
				...initialState,
			};
		case LOADING_SESSION_CART:
			return {
				...state,
				isLoading: true,
			};
		case LOAD_QUOTE_SUCCESS:
			return {
				...state,
				quoteId: action.quoteId,
			};
		case UPDATE_CART_ITEM_BOUNCE:
			return {
				...state,
				bounceFirstRowOnMount: action.payload.shouldBounce,
			};
		case GET_DELIVERY_DATES:
			delete state.cart.availableDeliveryDates;
			return { ...state };
		case GET_DELIVERY_DATES_SUCCESS:
			return {
				...state,
				cart: {
					...state.cart,
					availableDeliveryDates: action.payload,
				},
			};
		case SET_DELIVERY_DATE:
			return {
				...state,
				cart: {
					...state.cart,
					...action.payload,
				},
			};
		case GET_SESSION_CART_ERRORS: {
			return {
				...state,
				cart: {
					...state.cart,
					checkingErrors: true,
				},
			};
		}
		case SET_SESSION_CART_ERRORS: {
			return {
				...state,
				cart: {
					...state.cart,
					sessionCartErrors: action.payload,
					checkingErrors: false,
				},
			};
		}
		default:
			return {
				...state,
				isLoading: false,
			};
	}
}

export default cartReducer;
