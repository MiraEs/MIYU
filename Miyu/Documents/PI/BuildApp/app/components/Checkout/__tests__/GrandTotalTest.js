jest.mock('../../../store/configStore', () => ({}));
jest.mock('BuildLibrary');
jest.mock('BuildNative');
jest.mock('../../../lib/styles');
jest.mock('../../../lib/helpers', () => {
	return {
		subTotal: jest.fn(),
		calcGrandTotal: jest.fn(),
		toUSD: jest.fn(),
	};
});

jest.unmock('react-native');

import React from 'react';
import { shallow } from 'enzyme';
import GrandTotal from '../GrandTotal';
import cartReducer from '../../../../__mocks__/reducers/cartReducer';


const defaultProps = {
	cart: cartReducer.cart,
	storeCredit: 10,
	useStoreCredit: true,
};


describe('GrandTotal component', () => {
	const wrapper = shallow(<GrandTotal {...defaultProps} />);

	beforeEach(() => jest.resetModules());

	it('should render with default props', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
