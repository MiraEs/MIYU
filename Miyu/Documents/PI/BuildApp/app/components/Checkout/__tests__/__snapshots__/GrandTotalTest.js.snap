// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GrandTotal component should render with default props 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderColor": "#D0D3D4",
      "borderTopWidth": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
    }
  }
>
  <View
    accessibilityLabel="Total Price"
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
      }
    }
  >
    <Text
      lineHeight={false}
      size="large"
      style={
        Object {
          "flexGrow": 1,
        }
      }
      weight="bold"
    >
      Grand Total
    </Text>
    <Text
      color="primary"
      lineHeight={false}
      size="large"
      weight="bold"
    />
  </View>
</View>
`;
